crates/runtime/src/validation/mod.rs

```diff
@@ -1,53 +1,71 @@
-use wasmer_runtime_core::{
-    codegen::{Event, EventSink, FunctionMiddleware},
-    module::ModuleInfo,
-    wasmparser::Operator,
-};
+use svm_gas::{FuncPrice, ProgramVisitor};
+use wasmer::{ExternType, FunctionType};
+use wasmparser::Operator;
 
-use super::error::ParseError;
+pub fn calculate_gas_limit(wasm_module: &[u8], _gas_limit: u64) -> FuncPrice {
+    let program = svm_gas::read_program(wasm_module).unwrap();
+    let pricing_resolver = svm_gas::resolvers::ExampleResolver::default();
+    let program_pricing = svm_gas::ProgramPricing::new(pricing_resolver);
 
-/// The `ValidationMiddleware` has two main objectives:
-/// * validation - make sure the wasm is valid and doesn't contain and opcodes not supported by `svm` (for example: floats)
-/// * preprocessing - we want to know whether the input contains loops or not.
-///   In case there no loop we can later compute ahead-of-time the gas for each function,
-///   otherwise we'll have a dynamic gas-metering used.
-pub struct ValidationMiddleware;
+    let prices = program_pricing.visit(&program).unwrap();
+    prices
+}
 
-impl ValidationMiddleware {
-    pub fn new() -> Self {
-        Self {}
+/// Checks whether `wasm_module` exports a well-defined `svm_alloc` function.
+/// `svm_alloc` is required by SVM for all WASM code and must have a `I32 ->
+/// I32` type signature.
+pub fn validate_svm_alloc(wasm_module: &[u8]) -> bool {
+    let store = crate::new_store();
+    let module = if let Ok(m) = wasmer::Module::new(&store, wasm_module) {
+        m
+    } else {
+        return false;
+    };
+    for export in module.exports() {
+        let expected_type_signature =
+            ExternType::Function(FunctionType::new([wasmer::Type::I32], [wasmer::Type::I32]));
+        if export.name() == "svm_alloc" && export.ty() == &expected_type_signature {
```

Comment:
Maybe we can use the new `Exports` under `Program` (inside `svm_gas`) ?

Since it's not really a Gas-Specific logic. Maybe it will be a good idea to extract to a common crate.
(not sure about that)

Reply:
Shall I do it now? What do you think?

Reply:
Probably not, better to open a future issue...